File: src/PhysicsGame.java
// if anybody is reading this, I am sorry for the spahgetti code. I don't really want to make multiple files but you can if you want
import javax.swing.*;
import java.awt.*;
import java.awt.event.*;
import java.io.*;
import java.util.*;
import java.util.function.Consumer;
import java.util.List;
import java.util.ArrayList;

public class PhysicsGame {
    public static void main(String[] args) {
        SwingUtilities.invokeLater(PhysicsGame::createAndShowGUI);
    }

    private static void createAndShowGUI() {
        JFrame frame = new JFrame("Physics Game");
        frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        GamePanel gamePanel = new GamePanel();
        frame.setContentPane(gamePanel);
        frame.setSize(1000, 600);
        frame.setVisible(true);
    }
}

class GamePanel extends JPanel {
    private Circle circle;
    private ControlPanel controlPanel;

    public GamePanel() {
        setLayout(new BorderLayout());
        initializeComponents();
        configureLayout();
        startAnimation();
    }

    private void initializeComponents() {
        circle = new Circle();
        controlPanel = new ControlPanel(circle);
    }

    private void configureLayout() {
        add(circle, BorderLayout.CENTER);
        add(controlPanel, BorderLayout.SOUTH);
    }

    private void startAnimation() {
        new Thread(circle::animate).start();
    }
}

class ControlPanel extends JPanel {
    private final Circle circle;
    private JButton addBallButton;
    private JButton addMultiplierButton;
    private JButton saveProgressButton;
    private JLabel bounceLabel;
    private JLabel multiplierLabel;
    private JSlider slider;

    public ControlPanel(Circle circle) {
        this.circle = circle;
        initializeComponents();
        configureLayout();
        setupListeners();
    }

    private void initializeComponents() {
        slider = new JSlider(JSlider.HORIZONTAL, 1, 100, 10);
        addBallButton = new JButton("Add Ball (Cost: " + circle.getBounceThresholdForBall() + " Bounces)");
        addMultiplierButton = new JButton("Add Multiplier (Cost: " + circle.getBounceThresholdForMultiplier() + " Bounces)");
        saveProgressButton = new JButton("Save Progress");
        bounceLabel = new JLabel("Bounces: 0");
        multiplierLabel = new JLabel("Multiplier: 1");
    }

    private void configureLayout() {
        add(slider);
        add(addBallButton);
        add(addMultiplierButton);
        add(saveProgressButton);
        add(bounceLabel);
        add(multiplierLabel);
    }

    private void setupListeners() {
        slider.addChangeListener(e -> circle.setRadius(slider.getValue()));
        addBallButton.addActionListener(e -> circle.addBall());
        addMultiplierButton.addActionListener(e -> circle.purchaseMultiplier());
        saveProgressButton.addActionListener(e -> circle.saveProgress());
        circle.setBounceListener(count -> {
            bounceLabel.setText("Bounces: " + count);
            multiplierLabel.setText("Multiplier: " + circle.getMultiplier());
        });
    }
}

class Circle extends JPanel {
    public final int bounceThresholdForMultiplier = 100;
    public final int bounceThresholdForBall = 50;
    public final double friction = 0.99;
    public final double gravity = 0.2;
    public final double restitution = 0.6;
    public final List<Ball> balls = new ArrayList<>();
    public Ball draggedBall = null;
    public int bounceCount = 0;
    public int multiplier = 1;
    public Consumer<Integer> bounceListener;
    public Point lastMousePosition = new Point(0, 0);
    

    public void addBall() {
        if (bounceCount >= bounceThresholdForBall) {
            System.out.println("Adding ball");
            int newRadius = 10;
            int newX = (int) (Math.random() * (getWidth() - 2 * newRadius) + newRadius;
            int newY = (int) (Math.random() * (getHeight() - 2 * newRadius) + newRadius);
            balls.add(new Ball(newX, newY, newRadius));
            bounceCount -= bounceThresholdForBall;
            bounceListener.accept(bounceCount);
            repaint();
        } else {
            JOptionPane.showMessageDialog(this, "You need " +
                    (bounceThresholdForBall - bounceCount) + " more bounces to buy a ball!");
        }
    }

    public String getMultiplier() {
        return String.valueOf(multiplier);
    }

    public void purchaseMultiplier() {
        if (bounceCount >= bounceThresholdForMultiplier) {
            multiplier++;
            bounceCount -= bounceThresholdForMultiplier;
            bounceListener.accept(bounceCount);
            JOptionPane.showMessageDialog(this, "You now have a multiplier of " + multiplier + "!");
        } else {
            JOptionPane.showMessageDialog(this, "You need " +
                    (bounceThresholdForMultiplier - bounceCount) + " more bounces to buy a multiplier!");
        }
    }

    public int getBounceThresholdForMultiplier() {
        return bounceThresholdForMultiplier;
    }

    public String getBounceThresholdForBall() {
        return String.valueOf(bounceThresholdForBall);
    }

    public void setBounceListener(Consumer<Integer> listener) {
        this.bounceListener = listener;
    }

    public void setRadius(int newRadius) {
        for (Ball ball : balls) {
            ball.radius = newRadius;
        }
        repaint();
    }

    public void saveProgress() {
        try (BufferedWriter writer = new BufferedWriter(new FileWriter("game_progress.txt"))) {
            writer.write("Bounces: " + bounceCount);
            writer.newLine();
            writer.write("Multiplier: " + multiplier);
            writer.newLine();
            for (Ball ball : balls) {
                writer.write(String.format("Ball: x=%d, y=%d, radius=%d, vSpeed=%.2f, hSpeed=%.2f%n",
                        ball.x, ball.y, ball.radius, ball.verticalSpeed, ball.horizontalSpeed));
            }
            JOptionPane.showMessageDialog(this, "Progress saved successfully!");
        } catch (IOException ex) {
            JOptionPane.showMessageDialog(this, "Failed to save progress: " + ex.getMessage(), "Error", JOptionPane.ERROR_MESSAGE);
        }
    }

    @Override
    protected void paintComponent(Graphics g) {
        super.paintComponent(g);
        for (Ball ball : balls) {
            ball.paint(g);
        }
    }

    public void animate() {
        while (true) {
            try {
                Thread.sleep(10);
                move();
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
    }
    public void move() {
        for (Ball ball : balls) {
            ball.move();
        }
        repaint();
    }

    private void incrementBounceCount() {
        bounceCount += multiplier;
        bounceListener.accept(bounceCount);
    }
}

class Ball implements MouseListener, MouseMotionListener {
    public double x, y, radius, velocityX, velocityY;
    public long lastDragTime;
    public double horizontalSpeed;
    public double verticalSpeed;
    public boolean isDragging;

    public Ball(double x, double y, double radius) {
        this.x = x;
        this.y = y;
        this.radius = radius;
        this.velocityX = 0;
        this.velocityY = 0;
    }

    public boolean intersects(Ball other) {
        double dx = this.x - other.x;
        double dy = this.y - other.y;
        double distance = Math.sqrt(dx * dx + dy * dy);
        return distance < (this.radius + other.radius);
    }

    public double getX() {
        return this.x;
    }

    public double getY() {
        return this.y;
    }

    public double getRadius() {
        return this.radius;
    }

    public double getVelocityX() {
        return this.velocityX;
    }

    public double getVelocityY() {
        return this.velocityY;
    }

    public void setVelocityX(double velocityX) {
        this.velocityX = velocityX;
    }

    public void setVelocityY(double velocityY) {
        this.velocityY = velocityY;
    }

    public void move() {
        this.x += this.velocityX;
        this.y += this.velocityY;
    }

    public void paint(Graphics g) {
        g.setColor(getRandomColor());
        g.fillOval((int) (x - radius), (int) (y - radius), (int) (2 * radius), (int) (2 * radius));
    }

    public boolean contains(Point point) {
        double dx = point.x - x;
        double dy = point.y - y;
        return dx * dx + dy * dy <= radius * radius;
    }

    @Override
    public void mouseDragged(MouseEvent e) {
        Point lastDragPoint = e.getPoint();
        if (isDragging) {
            Point mousePoint = e.getPoint();
            double dx = mousePoint.x - lastDragPoint.x;
            double dy = mousePoint.y - lastDragPoint.y;

            x += dx;
            y += dy;

            lastDragPoint = mousePoint;
            lastDragTime = System.currentTimeMillis();
        }
    }

    
    @Override
    public void mouseClicked(MouseEvent e) {
    }

    @Override
    public void mousePressed(MouseEvent e) {
        if (contains(e.getPoint())) {
            isDragging = true;
            lastDragTime = System.currentTimeMillis();
        }
    }


    @Override
    public void mouseReleased(MouseEvent e) {
       
    }

    @Override
    public void mouseEntered(MouseEvent e) {
        // TODO Auto-generated method stub
    }

    @Override
    public void mouseExited(MouseEvent e) {
        // TODO Auto-generated method stub
    }

 

    @Override
    public void mouseMoved(MouseEvent e) {
        // TODO Auto-generated method stub
    }
    



    private Color getRandomColor() {
        Random random = new Random();
        return new Color(random.nextInt(256), random.nextInt(256), random.nextInt(256));
    }
}

File: src/PhysicsGame.kt
import javax.swing.*

object PhysicsGame {
    fun main(args: Array<String?>?) {
        SwingUtilities.invokeLater { createAndShowGUI() }
    }

    private fun createAndShowGUI() {    //creating the GUI
        val frame = JFrame("Physics Game")
        frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE)
        val gamePanel = GamePanel()
        frame.contentPane = gamePanel
        frame.setSize(1000, 600)
        frame.isVisible = true
    }
}

internal class GamePanel : JPanel() {
    //creating the game panel
    private var circle: Circle? = null
    private var controlPanel: ControlPanel? = null

    init {
        setLayout(BorderLayout())
        initializeComponents()
        configureLayout()
        startAnimation()
    }

    private fun initializeComponents() {
        circle = Circle()
        controlPanel = ControlPanel(circle!!)
    }

    private fun configureLayout() {
        add(circle, BorderLayout.CENTER)
        add(controlPanel, BorderLayout.SOUTH)
    }

    private fun startAnimation() {
        Thread { circle!!.animate() }.start()
    }
}

internal class ControlPanel(//creating the control panel
    private val circle: Circle
) : JPanel() {
    private var addBallButton: JButton? = null
    private var addMultiplierButton: JButton? = null
    private var saveProgressButton: JButton? = null
    private var bounceLabel: JLabel? = null
    private var multiplierLabel: JLabel? = null
    private var slider: JSlider? = null

    init {
        initializeComponents()
        configureLayout()
        setupListeners()
    }

    private fun initializeComponents() {   //initializing the components
        slider = JSlider(JSlider.HORIZONTAL, 1, 100, 10)
        addBallButton = JButton("Add Ball (Cost: " + circle.bounceThresholdForBall + " Bounces)")
        addMultiplierButton = JButton("Add Multiplier (Cost: " + circle.bounceThresholdForMultiplier + " Bounces)")
        saveProgressButton = JButton("Save Progress")
        bounceLabel = JLabel("Bounces: 0")
        multiplierLabel = JLabel("Multiplier: 1")
    }

    private fun configureLayout() {    //configuring the layout
        add(slider)
        add(addBallButton)
        add(addMultiplierButton)
        add(saveProgressButton)
        add(bounceLabel)
        add(multiplierLabel)
    }

    private fun setupListeners() {    //setting up the listeners
        slider.addChangeListener { e -> circle.setRadius(slider.getValue()) }
        addBallButton.addActionListener { e -> circle.addBall() }
        addMultiplierButton.addActionListener { e -> circle.purchaseMultiplier() }
        saveProgressButton.addActionListener { e -> circle.saveProgress() }
        circle.setBounceListener(Consumer<Integer> { count ->  // bounce listener
            bounceLabel.setText("Bounces: $count")
            multiplierLabel.setText("Multiplier: " + circle.multiplier)
        })
    }
}

internal class Circle : JPanel() {
    //creating the circles
    /*declaring variables used in physics and more*/
    val bounceThresholdForMultiplier = 100

    /*declaring variables used in physics and more*/
    val bounceThresholdForBall = 50

    /*declaring variables used in physics and more*/
    private val friction = 0.99

    /*declaring variables used in physics and more*/
    private val gravity = 0.2

    /*declaring variables used in physics and more*/
    private val restitution = 0.6

    /*declaring variables used in physics and more*/
    private val balls: List<Ball> = ArrayList()

    /*declaring variables used in physics and more*/
    private var draggedBall: Ball? = null

    /*declaring variables used in physics and more*/
    private var bounceCount = 0

    /*declaring variables used in physics and more*/
    var multiplier = 1
        private set

    /*declaring variables used in physics and more*/
    private var bounceListener: Consumer<Integer>? = null

    /*declaring variables used in physics and more*/
    private var lastMousePosition: Point = Point(0, 0)

    init {  //creating the circle
        setOpaque(false)
        balls.add(Ball(400, 300, 10))
        addMouseMotionListener(object : MouseMotionAdapter() {
            @Override
            fun mouseDragged(e: MouseEvent) {
                if (draggedBall != null) {
                    /*calculating the physics*/
                    val currentTime: Long = System.currentTimeMillis()
                    /*calculating the physics*/
                    val timeDiff = currentTime - draggedBall.lastDragTime
                    /*calculating the physics*/if (timeDiff > 0) {
                        /*calculating the physics*/
                        draggedBall.horizontalSpeed =
                            (e.getX() - lastMousePosition.x) / timeDiff.toDouble() * 50 // Speed scaling factor
                        /*calculating the physics*/draggedBall.verticalSpeed =
                            (e.getY() - lastMousePosition.y) / timeDiff.toDouble() * 50 // Speed scaling factor
                        /*calculating the physics*/
                    }
                    /*calculating the physics*/draggedBall.x = e.getX()
                    /*calculating the physics*/draggedBall.y = e.getY()
                    /*calculating the physics*/draggedBall.isDragging = true
                    /*calculating the physics*/lastMousePosition = e.getPoint()
                    /*calculating the physics*/draggedBall.lastDragTime = currentTime
                    repaint()
                }
            }
        })
        addMouseListener(object : MouseAdapter() {
            /*Looks for the mouse and if the mouse is on a ball, it gets dragged by the mouse.*/
            @Override
            fun mousePressed(e: MouseEvent) {
                for (ball in balls) {
                    if (ball.contains(e.getPoint())) {
                        draggedBall = ball
                        lastMousePosition = e.getPoint()
                        draggedBall.lastDragTime = System.currentTimeMillis()
                        break
                    }
                }
            }

            @Override
            fun mouseReleased(e: MouseEvent?) {  //if the mouse is released, the ball propells in the direction of the mouse
                if (draggedBall != null) {
                    draggedBall.isDragging = false
                    draggedBall = null
                }
            }
        })
    }

    fun purchaseMultiplier() {
        if (bounceCount >= bounceThresholdForMultiplier) {
            multiplier++
            bounceCount -= bounceThresholdForMultiplier
            bounceListener.accept(bounceCount)
        } else {
            JOptionPane.showMessageDialog(
                this, "You need " +
                        (bounceThresholdForMultiplier - bounceCount) + " more bounces to buy a multiplier!"
            )
        }
    }

    fun addBall() {
        if (bounceCount >= bounceThresholdForBall) {
            val newRadius = 10 // Default radius for new balls
            val newX = (Math.random() * (width - 2 * newRadius) + newRadius) as Int
            val newY = (Math.random() * (height - 2 * newRadius) + newRadius) as Int
            balls.add(Ball(newX, newY, newRadius))
            bounceCount -= bounceThresholdForBall
            bounceListener.accept(bounceCount)
            repaint()
        } else {
            JOptionPane.showMessageDialog(
                this, "You need " +
                        (bounceThresholdForBall - bounceCount) + " more bounces to buy a ball!"
            )
        }
    }

    fun setBounceListener(listener: Consumer<Integer?>) {
        bounceListener = listener
    }

    fun setRadius(newRadius: Int) {
        for (ball in balls) {
            ball.radius = newRadius
        }
        repaint()
    }

    fun saveProgress() {
        try {
            BufferedWriter(FileWriter("game_progress.txt")).use { writer ->
                writer.write("Bounces: $bounceCount")
                writer.newLine()
                writer.write("Multiplier: $multiplier")
                writer.newLine()
                for (ball in balls) {
                    writer.write(
                        String.format(
                            "Ball: x=%d, y=%d, radius=%d, vSpeed=%.2f, hSpeed=%.2f%n",
                            ball.x, ball.y, ball.radius, ball.verticalSpeed, ball.horizontalSpeed
                        )
                    )
                }
                JOptionPane.showMessageDialog(this, "Progress saved successfully!")
            }
        } catch (ex: IOException) {
            JOptionPane.showMessageDialog(
                this,
                "Failed to save progress: " + ex.getMessage(),
                "Error",
                JOptionPane.ERROR_MESSAGE
            )
        }
    }

    @Override
    protected fun paintComponent(g: Graphics) {
        super.paintComponent(g)
        for (ball in balls) {
            ball.paint(g)
        }
    }

    fun animate() {
        while (true) {
            try {
                Thread.sleep(10)
                move()
            } catch (e: InterruptedException) {
                e.printStackTrace()
            }
        }
    }

    fun move() {
        for (ball in balls) {
            ball.move()
        }
        repaint()
    }

    private fun incrementBounceCount() {
        bounceCount += multiplier
        bounceListener.accept(bounceCount)
    }

    internal inner class Ball(x: Int, y: Int, radius: Int) {
        var x: Int
        var y: Int
        var radius: Int
        var verticalSpeed = 0.0
        var horizontalSpeed = 0.0
        var isDragging = false
        var lastDragTime: Long = System.currentTimeMillis()
        var lastDragPoint: Point = Point(x, y)
        var ballColor: Color

        init {
            val random = Random() // used to generate random colors
            this.x = x
            this.y = y
            this.radius = radius
            lastDragPoint = Point(x, y)
            ballColor =
                Color(random.nextInt(256), random.nextInt(256), random.nextInt(256)) // random color for each new ball
        }

        operator fun contains(p: Point): Boolean {
            val dx: Double = p.x - x
            val dy: Double = p.y - y
            return dx * dx + dy * dy <= radius * radius
        }

        fun move() {
            if (isDragging) return
            x = (x + horizontalSpeed).toInt()
            y = (y + verticalSpeed).toInt()
            var bounced = false

            // collision with the bottom
            if (y + radius > height) {
                verticalSpeed *= -restitution
                y = height - radius
                horizontalSpeed *= friction
                bounced = true
            }

            // collision with the top
            if (y - radius < 0) {
                verticalSpeed *= -restitution
                y = radius
                horizontalSpeed *= friction
                bounced = true
            }

            // collision with the right
            if (x + radius > width) {
                horizontalSpeed *= -restitution
                x = width - radius
                verticalSpeed *= friction
                bounced = true
            }

            // collision with the left
            if (x - radius < 0) {
                horizontalSpeed *= -restitution
                x = radius
                verticalSpeed *= friction
                bounced = true
            }
            if (!bounced) {
                verticalSpeed += gravity // apply gravity if no collision
            } else {
                incrementBounceCount() // increment bounce count on collision
            }
        }

        fun paint(g: Graphics) {
            g.setColor(ballColor)
            g.fillOval(x - radius, y - radius, 2 * radius, 2 * radius)
        }
    }
}
